### ISSUE TITLE
Full support for local variable declarations

### ISSUE THREAD

**Issue by MirandaWood** [EYBlockchain/starlight#35]
While Starlight supports entirely secret or entirely public local variable declarations, it can lead to bugs in the output zApp when such variables interact with secret and public states.
Example of 'entirely public':
```
    function deposit(uint amount) public {
        bool hasBalance = erc20.transferFrom(msg.sender, address(this), amount);
        require(hasBalance == true);
        balances[msg.sender] += amount;
    }
```
Local var `hasBalance` only interacts with public states (`erc20` contract, `msg.sender`, contract address, and public param `amount`). The compiler copies this logic over into the shield contract and it all works.

However we may have something like:
```
    secret uint256 private x;
    secret mapping(uint256 => uint256) private y;

    function assign(secret uint256 param1, secret uint256 key, uint256 c) public {
        uint256 z = param1 + c;
        x = param1;
        unknown y[key] += z;
    }
```
The local variable `z` interacts with public param `c` and secret `param1` and `y`. Starlight doesn't (yet) know how to deal with this. 

---

A solution would be to introduce `LocalVariableIndicators` (alongside `StateVariableIndicators`), track the local variable's interactions with secret and public states, and generate code accordingly. It would also be simpler to use this method to give users warnings about mixing public/secret values, since the public value may reveal some/part of the secret.

With `z` as above, `indicators[z.id]` would have `interactsWithSecret: true` and `interactsWithPublic: true`. Since it `interactsWithSecret`, we must put those interactions in a zero-knowledge circuit - i.e. add
```
field z = param1 + c
```
to the circuit. 
Since it also `interactsWithPublic` (that public param being `c`) we must ensure that `c` is a public input to the circuit _as well as_ a public input to the shield contract. We check that the input `c` used in the circuit matches the one used on-chain as a parameter, without copying over the line including `param1`. (In a real world example, it may be needed to confirm the validity of some public value).

**Comment by MirandaWood** [comment]
Initial work has been done to implement `LocalVariableIndicators` on the branch #miranda/interactsWSecret.
This branch is a clone of #nfts (latest prelim traversal functionality).
The current order to be merged into master is:
`nfts` -> `interactsWSecret` -> incrementations fix #29

**Comment by MirandaWood** [comment]
Local variable indicators now merged, though code generation using those indicators is still in progress
